ngraph_bridge/ngraph_builder.cc

```diff
@@ -4725,6 +4725,101 @@ static Status TranslateStridedSliceOp(
   return Status::OK();
 }
 
+// For StridedSliceGradOp, see
+// .../tensorflow/core/kernels/strided_slice_op.cc
+// and .../tensorflow/core/api_def/base_api/api_def_StridedSliceGrad.pbtxt
+static Status TranslateStridedSliceGradOp(
+    const Node* op, const std::vector<const Tensor*>& static_input_map,
+    Builder::OpMap& ng_op_map) {
+  int tf_begin_mask;
+  TF_RETURN_IF_ERROR(GetNodeAttr(op->attrs(), "begin_mask", &tf_begin_mask));
+
+  int tf_end_mask;
+  TF_RETURN_IF_ERROR(GetNodeAttr(op->attrs(), "end_mask", &tf_end_mask));
+
+  int tf_new_axis_mask;
+  TF_RETURN_IF_ERROR(
+      GetNodeAttr(op->attrs(), "new_axis_mask", &tf_new_axis_mask));
+
+  int tf_shrink_axis_mask;
+  TF_RETURN_IF_ERROR(
+      GetNodeAttr(op->attrs(), "shrink_axis_mask", &tf_shrink_axis_mask));
+
+  int tf_ellipsis_mask;
+  TF_RETURN_IF_ERROR(
+      GetNodeAttr(op->attrs(), "ellipsis_mask", &tf_ellipsis_mask));
+
+  NGRAPH_VLOG(5) << "Arguments to make_slice_plan: "
+                 << "begin mask: " << tf_begin_mask
+                 << ", end mask: " << tf_end_mask
+                 << ", new axis mask: " << tf_new_axis_mask
+                 << ", shrink axis mask: " << tf_shrink_axis_mask
+                 << ", ellipsis mask: " << tf_ellipsis_mask;
+
+  // get original shape
+  std::vector<int64> original_shape;
+  TF_RETURN_IF_ERROR(
+      GetStaticInputVector(op, 0, static_input_map, &original_shape));
+  ng::Shape ng_original_shape(
+      std::vector<size_t>(original_shape.begin(), original_shape.end()));
+  NGRAPH_VLOG(5) << "Original shape: " << ng::join(ng_original_shape);
+
+  shared_ptr<ng::Node> ng_delta;
+  TF_RETURN_IF_ERROR(GetInputNode(ng_op_map, op, 4, &ng_delta));
+
+  // get begin, end, and stride
+  std::vector<int64> begin_vec;
+  TF_RETURN_IF_ERROR(GetStaticInputVector(op, 1, static_input_map, &begin_vec));
+  std::vector<int64_t> begin_vec_longint(begin_vec.begin(), begin_vec.end());
+
+  std::vector<int64> end_vec;
+  TF_RETURN_IF_ERROR(GetStaticInputVector(op, 2, static_input_map, &end_vec));
+  std::vector<int64_t> end_vec_longint(end_vec.begin(), end_vec.end());
+
+  std::vector<int64> stride_vec;
+  TF_RETURN_IF_ERROR(
+      GetStaticInputVector(op, 3, static_input_map, &stride_vec));
+  std::vector<int64_t> stride_vec_longint(stride_vec.begin(), stride_vec.end());
+
+  // get slice plan
+  auto convert_mask_to_axes = [](const int mask) {
+    ng::AxisSet axes{};
+    for (auto i = 0; i < sizeof(int) * 8; ++i) {
+      if ((unsigned char)(mask >> i & 0x01) == 1) {
```

Comment:
Minor: Please add brackets to make the order of operation more readable.

Reply:
A lot of the code of TranslateStridedSlice and TranslateStridedSliceGrad is similar (for example convert_mask_to_axes). Can we use a single function or a helper function or somehow reuse the code?

Reply:
agree